sdk/src/main/scala/com/horizen/account/state/AccountStateView.scala

```diff
@@ -0,0 +1,330 @@
+   */
+  override def applyTransaction(
+      tx: SidechainTypes#SCAT,
+      txIndex: Int,
+      blockGasPool: GasPool,
+      blockContext: BlockContext,
+      finalizeChanges: Boolean = true
+  ): Try[EthereumConsensusDataReceipt] = Try {
+    if (!tx.isInstanceOf[EthereumTransaction])
+      throw new IllegalArgumentException(s"Unsupported transaction type ${tx.getClass.getName}")
+
+    val ethTx = tx.asInstanceOf[EthereumTransaction]
+    val txHash = BytesUtils.fromHexString(ethTx.id)
+    val msg = ethTx.asMessage(blockContext.baseFee)
+
+    // Tx context for stateDB, to know where to keep EvmLogs
+    setupTxContext(txHash, txIndex)
+
+    log.debug(s"applying msg: used pool gas ${blockGasPool.getUsedGas}")
+    // apply message to state
+    val status = try {
+      applyMessage(msg, blockGasPool, blockContext)
+      ReceiptStatus.SUCCESSFUL
+    } catch {
+      // any other exception will bubble up and invalidate the block
+      case err: ExecutionFailedException =>
+        log.error(s"applying message failed, tx.id=${ethTx.id}", err)
+        ReceiptStatus.FAILED
+    } finally {
+      // finalize pending changes, clear the journal and reset refund counter
+      if (finalizeChanges)
+        stateDb.finalizeChanges()
+    }
+    val consensusDataReceipt = new EthereumConsensusDataReceipt(
+      ethTx.version(), status.id, blockGasPool.getUsedGas, getLogs(txHash))
+    log.debug(s"Returning consensus data receipt: ${consensusDataReceipt.toString()}")
+    log.debug(s"applied msg: used pool gas ${blockGasPool.getUsedGas}")
+
+    consensusDataReceipt
+  }
+
+  override def isEoaAccount(address: Array[Byte]): Boolean = stateDb.isEoaAccount(address)
+
+  override def isSmartContractAccount(address: Array[Byte]): Boolean = stateDb.isSmartContractAccount(address)
+
+  override def accountExists(address: Array[Byte]): Boolean = !stateDb.isEmpty(address)
+
+  // account modifiers:
+  override def addAccount(address: Array[Byte], code: Array[Byte]): Unit =
+    stateDb.setCode(address, code)
+
+  override def increaseNonce(address: Array[Byte]): Unit =
+    stateDb.setNonce(address, getNonce(address).add(BigInteger.ONE))
+
+  @throws(classOf[ExecutionFailedException])
+  override def addBalance(address: Array[Byte], amount: BigInteger): Unit = {
+    amount.signum() match {
+      case x if x == 0 => // amount is zero
+      case x if x < 0 =>
+        throw new ExecutionFailedException("cannot add negative amount to balance")
+      case _ =>
+        log.debug(s"Adding $amount to addr ${BytesUtils.toHexString(address)}")
+        stateDb.addBalance(address, amount)
+    }
+  }
+
+  @throws(classOf[ExecutionFailedException])
+  override def subBalance(address: Array[Byte], amount: BigInteger): Unit = {
+    // stateDb lib does not do any sanity check, and negative balances might arise (and java/go json IF does not correctly handle it)
+    // TODO: for the time being do the checks here, later they will be done in the caller stack
```

Comment:
Is this true? I vote for keeping it here

Reply:
we do it now during tx validation.
let's remove TODO and the check `case x if x > 0 && stateDb.getBalance(address).compareTo(amount) < 0 =>` below. It consts too much